### PR TITLE
Calcite Index Cost

### PR DESCRIPTION
**Pull request by alexeevm**
HOW WE COST INDEXES
        Since Volt only supports TREE based indexes the cost can be approximated as following
```
log(N / Sparsity)  + Sparsity  - 1
```
        where N is number of rows in a table and Sparsity represents the number of identical rows -
        Sparsity is 10 if 1 out 10 rows is a duplicate.  For a UNIQUE index Sparcity is 1
        To account for a multicolumn indexes and partial coverage, Sparsity is adjusted to be
```
        Adjusted Sparsity = Sparsity ** (Index Column Count - Key Width + 1)
```
        where the Index Column Count denotes number of columns / expressions in the index
        and Key Width denotes the number of Index columns / expressions covered by a given access path
         "Covering cell" indexes get a special adjustment to make them look more favorable
        Partial Indexes still needs to be discounted further

### CODE REVIEW

File: src/frontend/org/voltdb/plannerv2/rel/physical/VoltPhysicalSort.java [hunk 1]
```diff
@@ -111,13 +111,16 @@ public RelWriter explainTerms(RelWriter pw) {
...
+    public double estimateRowCount(RelMetadataQuery mq) {
+        return getInput(0).estimateRowCount(mq);
+     }
 
-        RelOptCost defaultCost = super.computeSelfCost(planner, mq);
-        return planner.getCostFactory().makeCost(rowCount, defaultCost.getCpu(), defaultCost.getIo());
+    @Override
+    public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        double rowCount = estimateRowCount(mq);
+        //the worst-case time complexity is mandated to be O(nlogn)
```
[the remark this answers is not in the document]
Comment (alexeevm) [reply src/frontend/org/voltdb/plannerv2/rel/physical/VoltPhysicalSort.java]: You mean if VoltPhysicalSort sits on top of a VoltPhysicalTableSequentialScan and there is an index that satisfies the sort order, right? In this case VoltPhysicalSort / VoltPhysicalTableSequentialScan will be transformed into a VoltPhysicalTableIndexScan. Both plans (one with VoltPhysicalSort / VoltPhysicalTableSequentialScan and another one with just a VoltPhysicalTableIndexScan) will compete for a minimal cost ( n + nlog(n) vs log(n)) and the plan with just an index scan should win.
Besides, you can't even check from within a VoltPhysicalScan whether its collation satisfies a possible index scan - VoltPhysicalScan's child is a RelSet (and not a VoltPhysicalTableSequentialScan). It's only the VoltPSortScanToIndexRule that matches both nodes can make a determination

File: src/frontend/org/voltdb/plannerv2/rel/physical/VoltPhysicalAggregate.java [hunk 2]
```diff
@@ -107,10 +108,29 @@ protected String computeDigest() {
...
+        // Assume that each GROUP BY column has 50% of the value count.
+        // Therefore one GROUP BY column has .5 * rowCount,
+        // 2 sort columns give .75 * rowCount.
+        // Zero GROUP BY columns yields 1 row (or 0 if the input is empty).
+        final int groupCount = groupSet.cardinality();
+        if (groupCount == 0) {
+            return 1;
+        } else {
+            double childRowCount = getInput(0).estimateRowCount(mq);
+            double rowCount = childRowCount * 1.0 - Math.pow(.5, groupCount);
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/plannerv2/rel/physical/VoltPhysicalAggregate.java]: Missing a parenthesis here?

**Comment by kaikaizi** [comment]
Hi Mike,
Would you please hold on to your updates for now? I am going to create a new PR from VoltDB repo (rather than your forked repo) today, and you could resume your work from that branch.

**Comment by alexeevm** [comment]
Lukai,

Sure,  will wait for the new PR. I am still working on index selection - adopting the existing TestIndexSelection into the Calcite.

**Pull request closed without merging**